solver/src/driver.rs

```diff
@@ -104,91 +105,57 @@ impl Driver {
+            let metrics = &self.metrics;
+            async move {
+                let start_time = Instant::now();
+                let settlement = solver.solve(liquidity, gas_price).await;
+                metrics.settlement_computed(solver.to_string().as_str(), start_time);
+                (index, settlement)
+            }
+        }))
+        .await
+        .into_iter()
+        .filter_map(|(index, settlement)| {
+            let settlement = match settlement {
+                Ok(settlement) => settlement,
                 Err(err) => {
-                    tracing::error!("solver {} error: {:?}", solver, err);
-                    None
+                    tracing::error!("solver {} error: {:?}", self.solver[index].to_string(), err);
+                    return None;
                 }
+            }?;
+            if settlement.trades().is_empty() {
+                return None;
+            }
+            let objective_value = settlement.objective_value(prices);
+            Some(SolverSettlement {
+                index,
+                settlement,
+                objective_value,
             })
-            .collect();
-        for (solver, settlement) in settlements.iter() {
-            tracing::info!(
-                "{} found solution with objective value: {}",
-                solver,
-                settlement.objective_value(&estimated_prices)
-            );
-        }
+        })
+        .collect()
+    }
 
-        // Sort by key in descending order
-        settlements.sort_by_cached_key(|(_, settlement)| {
-            Reverse(settlement.objective_value(&estimated_prices))
-        });
+    // Go through the settlements in order until the first success.
+    async fn submit_settlements(&self, settlements: Vec<SolverSettlement>) {
         let settle_orders_older_than =
             chrono::offset::Utc::now() - chrono::Duration::from_std(self.min_order_age).unwrap();
-        for (solver, settlement) in settlements {
-            tracing::info!("{} computed {:?}", solver, settlement);
-
-            if settlement.trades().is_empty() {
-                tracing::info!("Skipping empty settlement");
-                continue;
-            }
-
+        for settlement in settlements {
             // If all orders are younger than self.min_order_age skip settlement. Orders will still
             // be settled once they have been in the order book for longer. This makes coincidence
             // of wants more likely.
-            let should_be_settled_immediately = settlement
-                .trades()
-                .iter()
-                .any(|trade| trade.order.order_meta_data.creation_date <= settle_orders_older_than);
+            let should_be_settled_immediately =
+                settlement.settlement.trades().iter().any(|trade| {
+                    trade.order.order_meta_data.creation_date <= settle_orders_older_than
+                });
             if !should_be_settled_immediately {
```

Comment:
Maybe all settlement filtering should happen in one place? Right now we filter:
- settlements with no trades
- settlements with no "old enough" orders

in two separate places. I think, it would be nice to have a `should_settle` function (with a better name :stuck_out_tongue:) that does this.

Reply:
Also, for example, there is no need to compute the objective value of a settlement with "too recent" orders.

Reply:
One reason for this is that this check is time dependent so it has to come after anything async like running the solver. But yes this part could be improved more.